File: app/assets/javascripts/models/serialize.js
AQ.Plan.record_methods.serialize = function() {

  var plan = this, ops;

  p = {
    id: plan.id,
    operations: aq.collect(plan.operations, op => op.serialize() ),
    wires: aq.collect(plan.wires, w => w.serialize() ),
    user_budget_association: plan.uba,
    status: plan.status,
    cost_limit: plan.cost_limit,
    name: plan.name,
    rid: plan.rid,
    layout: plan.serialize_module(plan.base_module)
  }

  return p;

}

AQ.Plan.record_methods.serialize_module = function(m) {

  var plan = this,
      props = [ "id", "parent_id", "name", "x", "y", "width", "height", "model", "input", "output", "documentation" ],
      sm = {};

  for ( var p in props ) {
    sm[props[p]] = m[props[p]];
  }

  sm.input  = aq.collect(m.input, i => plan.serialize_module_io(i));
  sm.output = aq.collect(m.output, o => plan.serialize_module_io(o));  

  sm.children = aq.collect(m.children, c => plan.serialize_module(c));
  sm.wires = aq.collect(m.wires, w => w.serialize())

  return sm;

}

AQ.Plan.record_methods.serialize_module_io = function(io) {

  var plan = this,
      props = [ "id", "x", "y", "width", "height", "model" ],
      sio = {};

  for ( var p in props ) {
    sio[props[p]] = io[props[p]];
  }  
  
  return sio;

}

AQ.Operation.record_methods.serialize = function() {

  var op = this;

  return {
    id: op.id,
    operation_type_id: op.operation_type_id,
    field_values: aq.collect(op.field_values, fv => {
      var efv = fv.serialize();
      efv.allowable_field_type_id = op.form[fv.role][fv.name] && op.form[fv.role][fv.name].aft ? 
                                    op.form[fv.role][fv.name].aft.id :
                                    null;
      if ( !fv.field_type.array && fv.routing ) {
        op.assign_sample(efv,op.routing[fv.routing])
      }
      return efv;
    }),
    status: op.status,
    user_id: op.user_id,
    x: op.x,
    y: op.y,
    routing: op.routing,
    rid: op.rid,
    parent_id: op.parent_id
  };

}

AQ.FieldValue.record_methods.serialize = function() {

  var fv = this,
      props = [ "id", "name", "child_item_id",, "child_sample_id", "value", "role", 
                "field_type_id", "item", "row", "column", "row", "column",
                "parent_class", "parent_id", "routing", "rid" ],
      efv = {};

  aq.each(props, p => efv[p] = fv[p]); 


  if ( fv.field_type.array ) {
    efv.array = true;
  }

  return efv;

}

AQ.Wire.record_methods.serialize = function() {

  var w = this;

  return {
    id: w.id,
    from_id: w.from_id,
    to_id: w.to_id,
    from: { rid: this.from.rid },
    to: { rid: this.to.rid },
    active: w.active,
    rid: w.rid
  }

}

File: app/planning/serialize.rb
module Serialize

  def self.serialize_fv(fv)

    rfv = {}
    FieldValue.attribute_names.each { |name| rfv[name] = fv[name] }
    rfv

  end

  def self.serialize(plan)

    ops = plan.operations

    op_ids = ops.collect { |o| o.id }
    ot_ids = ops.collect { |o| o.operation_type_id }    

    job_associations = JobAssociation.includes(:job).where(operation_id: op_ids).collect { |ja| ja.as_json(include: :job) }
    operation_types = OperationType.where(id: ops.collect { |o| o['operation_type_id'] }).as_json
    field_types = FieldType.includes(allowable_field_types: %i[sample_type object_type])
                           .where(parent_class: 'OperationType', parent_id: ot_ids)
                           .collect(&:as_json)
    operation_types.each do |ot|
      ot['field_types'] = field_types.select { |ft| ft['parent_id'] == ot['id'] }
    end

    field_values = FieldValue
                   .where(parent_class: 'Operation', parent_id: op_ids)
                   .collect { |fv| serialize_fv fv }

    fids = field_values.collect { |fv| fv['id'] }

    wires = Wire.where('from_id in (?) OR to_id in (?)', fids, fids)

    field_values = self.complete_field_values(field_values, field_types)

    sops = ops.as_json

    sops.each do |op|
      op['field_values'] = field_values.select { |fv| fv['parent_id'] == op['id'] }
      op['job_associations'] = job_associations.select { |ja| ja['operation_id'] == op['id'] }
    end

    {
      id: plan.id,
      name: plan.name,
      budget_id: plan.budget_id,
      folder: plan.folder,
      user_id: plan.user_id,
      created_at: plan.created_at,
      updated_at: plan.updated_at,
      status: plan.status,
      operations: sops,
      wires: wires,
      layout: plan.layout
    }

  end

  def self.get_aft_for fv, afts
    aft_list = afts.select { |aft| aft.id == fv["allowable_field_type_id"]  }
    if aft_list.length > 0 
      aft_list[0]
    else 
      nil
    end
  end

  def self.complete_field_values field_values, field_types

    fvs = field_values

    fvs.each do |fv|

      fts = field_types.select { |ft| ft['id'] == fv['field_type_id'] }
      fv['field_type'] = fts[0] if fts.length == 1

    end

    fvs

  end

  def self.fast_operation_types(dep_only = true)

    ots = OperationType
    ots = ots.where(deployed: true) if dep_only
    ot_ids = ots.collect(&:id)
    fts = FieldType.includes(:allowable_field_types).where(parent_class: 'OperationType', parent_id: ot_ids)
    st_ids = fts.collect { |ft| ft.allowable_field_types.collect(&:sample_type_id) }.flatten
    sts = SampleType.where(id: st_ids)
    ob_ids = fts.collect { |ft| ft.allowable_field_types.collect(&:object_type_id) }.flatten
    obs = ObjectType.where(id: ob_ids)

    ots.collect do |ot|

      sot = ot.as_json

      sot[:field_types] = fts.select { |ft| ft.parent_id == ot.id }
                             .collect do |ft|
        sft = ft.as_json plain: true
        sft[:allowable_field_types] = ft.allowable_field_types.collect do |aft|
          {
            id: aft.id,
            sample_type_id: aft.sample_type_id,
            object_type_id: aft.object_type_id,
            field_type_id: aft.field_type_id,
            sample_type: sts.find { |st| st.id == aft.sample_type_id }.as_json,
            object_type: obs.find { |object_type| object_type.id == aft.object_type_id }.as_json
          }
        end
        sft
      end

      sot

    end

  end

  def self.item_history(item)

    fvs = FieldValue.where(parent_class: 'Operation', child_item_id: item.id)
    op_ids = fvs.collect(&:parent_id)
    ops = Operation.includes(:jobs, :operation_type, :plan_associations).where(id: op_ids)

    history = fvs.collect do |fv|
      {
        field_value: fv,
        operation: ops.find { |op| op.id == fv.parent_id }.as_json(include: [:plan_associations, :operation_type, { jobs: { except: :state } }])
      }
    end
    history.select { |h| h[:operation] }

  end

end
